### PR TITLE
Runner - Fetch benchmarks results on all nodes

### PR DESCRIPTION
**Pull request by abuccts**
Fetch benchmarks results on all nodes, will rsync after each benchmark.
The results directory structure on control node is as follows:

```
outputs/
└── datetime
    ├── nodes
    │   └── node-0
    │       ├── benchmarks
    │       │   ├── benchmark-0
    │       │   │   ├── rank-0
    │       │   │   │   └── results.json
    │       └── sb-exec.log
    ├── sb-run.log
    └── sb.config.yaml
```

### CODE REVIEW

File: superbench/runner/playbooks/fetch_results.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+- name: Fetch Results
+  hosts: all
+  gather_facts: true
+  vars:
+    workspace: '{{ ansible_user_dir }}/sb-workspace'
+  tasks:
+    - name: Synchronize Output Directory
+      ansible.posix.synchronize:
```
Comment (abuccts) [inline superbench/runner/playbooks/fetch_results.yaml:8]: add dep

File: superbench/runner/playbooks/fetch_results.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+        mode: pull
+        src: '{{ workspace }}/{{ sb_output_dir }}/'
+        dest: '{{ absolute_output_dir }}/nodes/{{ ansible_hostname }}'
```
Comment (cp5555) [inline superbench/runner/playbooks/fetch_results.yaml:11]: In future, we should think about uploading all results to storage directly.

File: superbench/executor/executor.py [hunk 3]
```diff
@@ -122,21 +123,34 @@ def __exec_benchmark(self, context, log_suffix):
...
+            logger.warning('Benchmark output root directory %s is not empty.', str(benchmark_output_root_dir))
             for i in itertools.count(start=1):
-                backup_dir = benchmark_output_dir.with_name('{}.{}'.format(benchmark_name, i))
+                backup_dir = benchmark_output_root_dir.with_name('{}.{}'.format(benchmark_name, i))
                 if not backup_dir.is_dir():
-                    benchmark_output_dir.rename(backup_dir)
+                    benchmark_output_root_dir.rename(backup_dir)
                     break
-        benchmark_output_dir.mkdir(mode=0o755, parents=True, exist_ok=True)
+        self.__get_benchmark_dir(benchmark_name).mkdir(mode=0o755, parents=True, exist_ok=True)
```
Comment (cp5555) [inline superbench/executor/executor.py:153]: I think we should consider the failure if we cannot create or open this dir.

File: superbench/runner/runner.py [hunk 4]
```diff
@@ -156,6 +157,19 @@ def check_env(self):    # pragma: no cover
             )
         )
 
+    def fetch_results(self):    # pragma: no cover
+        """Fetch benchmark results on all nodes."""
+        (self._output_path / 'nodes').mkdir(mode=0o755, parents=True, exist_ok=True)
```
Comment (cp5555) [inline superbench/runner/runner.py:162]: For example, the disk is full, how to handle it.

File: superbench/executor/executor.py [hunk 5]
```diff
@@ -122,21 +123,38 @@ def __exec_benchmark(self, context, log_suffix):
...
         return None
 
+    def __get_benchmark_dir(self, benchmark_name):
+        """Get output directory for benchmark's current rank.
+
+        Args:
+            benchmark_name (str): Benchmark name.
+        """
+        benchmark_output_dir = self._output_path / 'benchmarks' / benchmark_name
+        for rank_env in ['PROC_RANK', 'LOCAL_RANK']:
```
Comment (guoshzhao) [inline superbench/executor/executor.py:133]: Can we say rank value match the gpu id for both micro and model benchmark? In other words, how can we match the log to hardware?

File: superbench/executor/executor.py [hunk 6]
```diff
@@ -122,21 +123,38 @@ def __exec_benchmark(self, context, log_suffix):
...
-        benchmark_output_dir.mkdir(mode=0o755, parents=True, exist_ok=True)
+        benchmark_output_root_dir = self._output_path / 'benchmarks' / benchmark_name
+        try:
+            if benchmark_output_root_dir.is_dir() and any(benchmark_output_root_dir.iterdir()):
+                logger.warning('Benchmark output root directory %s is not empty.', str(benchmark_output_root_dir))
+                for i in itertools.count(start=1):
+                    backup_dir = benchmark_output_root_dir.with_name('{}.{}'.format(benchmark_name, i))
+                    if not backup_dir.is_dir():
+                        benchmark_output_root_dir.rename(backup_dir)
+                        break
```
Comment (guoshzhao) [inline superbench/executor/executor.py:153]: Can we make it as a util function?  Looks the input can be root_dir and sub_dir.

**Review by cp5555 — APPROVED** [review]

File: superbench/executor/executor.py (hunk 5 above)
Comment (abuccts) [reply superbench/executor/executor.py:133]: it's decided by benchmark and its config, e.g., you can use gpu_id=rank/2, or gpu_id=rank*2, etc. for pytorch ddp, gpu_id=rank.

File: superbench/executor/executor.py (hunk 6 above)
Comment (abuccts) [reply superbench/executor/executor.py:153]: do you mean `__get_benchmark_dir`? the rotation here only handles root dir.

Comment (guoshzhao) [reply superbench/executor/executor.py:153]: I mean the function to check the existence of one dir and rename it. We have the similiar function in sb v0.1

Comment (abuccts) [reply superbench/executor/executor.py]: updated

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**